### DISCUSSION TITLE
Next 13.2.1 – reading the body of a POST request in API routes

### DISCUSSION THREAD

**Discussion by sebpowell** [vercel/next.js#46483] [Help] [answered]
I've been working on upgrading our website to use Next 13. We have some API routes located within the `app` folder that look something like this:

```TS

// /app
//  /api
//   /test
//    /route.ts

type Payload = {
  value: string;
}

export async function POST(req: NextRequest) {
  
  const data: Payload = req.body;
  
  // This doesn't work
  const { value } = data;

  return NextResponse.json({});
}
```

It looks like `req.body` is now a readable stream, and I'm not sure how to go about actually reading the JSON payload that comes through. 

I believe this wasn't the case in the previous version, and I couldn't find anything in the upgrade guide that details how to handle this.

Any help would be greatly appreciated!

**Comment by icyJoseph** [comment, marked as answer]
You just need to invoke the `.json()` method and await it.

```ts
export async function POST(req: Request) {
  const data = await req.json();
  console.log(data);
  return new Response("Hello, Next.js!");
}
```

How do I know this? The Request object is part of the Web API: https://developer.mozilla.org/en-US/docs/Web/API/Request/json

**Reply by sebpowell** [reply]
@icyJoseph – ahhh of course. That's perfect, thank you!

**Reply by BhargavEverestek** [reply]
![image](https://github.com/vercel/next.js/assets/43741543/a043c416-4c86-4515-938f-1100523a9969)

`request.json()` throws an error. using Next 13.4,4 tried 13.3.3 as well

**Reply by icyJoseph** [reply]
Are you perhaps trying to statically optimise the route? Like doing `export const dynamic = "error"`

**Reply by SebasCC99** [reply]
Thank you so much!

**Reply by AustinBajka** [reply]
Was really hoping this would work for me but even after following the documentation on this (which is just what you put) from The Next.JS docs, I'm still running into an error. I'm not sure how to get this any simpler.

Button that is called in the page.tsx

`const sendHelloFunction = async () => {
  const response = await fetch('/api/hello', {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      body: JSON.stringify({
        message: "hello",
      }),
    },
  })
}

const SendHelloButton = () => {
  return (
    <form action={sendHelloFunction}>
    <button type="submit"> Test Send Hello Button</button>
    </form>
  )
}

export default SendHelloButton;`


And then the route to the route handler or route.ts

/src/app/api/hello/route.ts

`import { NextRequest, NextResponse } from "next/server";

export async function GET() {
  return NextResponse.json("Hello, Next.js");
}

export async function POST(request: NextRequest) {
  const body = await request.json();
  console.log("🚀 ~ file: route.ts:9 ~ PUT ~ body:", body);
  return NextResponse.json("Hello, Next.js");
}`

And I'm getting this error from the route handler route when I try to change the request to JSON.

`- error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (/Users/austinwhitbeck/route-handler-test/node_modules/next/dist/compiled/undici/index.js:2:4905)
    at successSteps (/Users/austinwhitbeck/route-handler-test/node_modules/next/dist/compiled/undici/index.js:2:4473)
    at /Users/austinwhitbeck/route-handler-test/node_modules/next/dist/compiled/undici/index.js:2:65581
    at node:internal/process/task_queues:141:7
    at AsyncResource.runInAsyncScope (node:async_hooks:201:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:138:8)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)`
    
    
    Any insight as to why this might be? I don't think I could be making this any simpler. And I'm referencing the next docs from here:
    
    https://nextjs.org/docs/app/building-your-application/routing/route-handlers
    
    and their example code:
    
    
![image](https://github.com/vercel/next.js/assets/77368029/2ab28361-e187-42a4-9840-5a3b418b6869)

**Reply by faaiz99** [reply]
This happens because by default Next form data is sent through form data x-www-form-urlencoded format instead of the common json format and the data received is in another encoding so it cannot parse JSON from it. Next docs suggests adding custom headers to change it to  application/json but they do not work with next:export. None of the solutions described above work. So we're kinda stuck with this now. 

You can verify the issue by sending a post request through postman using x-www-form-urlencoded format. 

I worked around by directly using supabase inside the server component.

**Reply by rizvan555** [reply]
Thanks icyJoseph, I couldn't find the way, you helped me .....

**Reply by miloskapovic** [reply]
You can read request body like this for NextRequest
async function streamToString(stream: any) {
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString('utf8');
}

const data = streamToString(request.body)

**Reply by AbhigyanBafna** [reply]
Thanks @miloskapovic After searching for an hour, your solution finally worked.

**Reply by vitaliiurban** [reply]
But for some reason it didn't help me, I got the following error "src\app\api\customers\page.ts (7:29) @ json
 ⨯ TypeError: request.json is not a function
    at POST (./src/app/api/customers/page.ts:6:32)
    at stringify (<anonymous>)
digest: "49170738""

`export default async function POST(request: Request) {`
  `const data = await request.json();`
 ` console.log(data);`
 ` return new Response(JSON.stringify(data), {`
`    headers: { "content-type": "application/json" },`
`    status: 201,`
 ` });`

![image](https://github.com/vercel/next.js/assets/118912994/57708900-0c45-4725-9ae6-e78feac7a0d1)

**Reply by icyJoseph** [reply]
> But for some reason it didn't help me, I got the following error "src\app\api\customers\page.ts (7:29) @ json
>  ⨯ TypeError: request.json is not a function
>     at POST (./src/app/api/customers/page.ts:6:32)
>     at stringify (<anonymous>)
> digest: "49170738""
> 
> `export default async function POST(request: Request) {`
>   `const data = await request.json();`
>  ` console.log(data);`
>  ` return new Response(JSON.stringify(data), {`
> `    headers: { "content-type": "application/json" },`
> `    status: 201,`
>  ` });`
> 
> ![image](https://github.com/vercel/next.js/assets/118912994/57708900-0c45-4725-9ae6-e78feac7a0d1)
> 

You have a `page.ts`, change it to `route.ts`

**Reply by 14790897** [reply]
thank you

**Comment by vladislav-gorovenko** [comment]
```
export async function DELETE(request: Request) {
  const { id }: Partial<Todo> = await request.json();

  if (!id) return NextResponse.json({ deleted: false });

  const res = await fetch(`${DATA_SOURCE_URL}/${id}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      "API-Key": API_KEY,
    },
  });

  const data = await res.json();

  return NextResponse.json({ ...data, deleted: true });
}
```
<img width="1094" alt="Screenshot 2023-06-09 at 11 10 07" src="https://github.com/vercel/next.js/assets/38355614/d7b27c88-0d3d-4089-ba63-3b86b1fdec57">

Error message: `error SyntaxError: Unexpected end of JSON input at JSON.parse`

For some reason can make it to work :(

**Reply by icyJoseph** [reply]
Nvm, it is just that DELETE has weird body semantics/expectations/specification -

**Reply by adrivelasco** [reply]
same issue here, how did you solve it?

**Reply by adrivelasco** [reply]
actually I can ignore the request body if the http method is a `DELETE` but just asking.

```ts
const data = method !== 'DELETE' ? await req.json() : undefined;
```

**Reply by snowden-fu** [reply]
can i ask what http test tool are you using?

**Reply by faaiz99** [reply]
that's thunder client

**Comment by maxyou** [comment]
this work for me: https://stackoverflow.com/a/76543092/2468360

const body = await req.text();

**Reply by jdamon96** [reply]
doesn't fix for me

**Reply by Rasukarusan** [reply]
thanks this work!

**Reply by aditodkar** [reply]
@maxyou Which nextJS version are you using? Please share `next` package version from package.json.

**Reply by maxyou** [reply]
"next": "13.4.1",
https://github.com/maxyou/nextjs13-starter

@aditodkar

**Comment by jdamon96** [comment]
I'm currently having this issue with Next.js v13.4.9

I call my API route in a form component with the following code:
```
const onSubmit = () => {
    const data = {
      field: <value>,
    };

    const api_url = process.env.URL + "/app/api";

    const api_req_options = {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify(data),
    };

    const res = await fetch(api_url, api_req_options);
}
```

Where this is my route handler code:
```
// /app/api/route.ts
import { NextRequest, NextResponse } from "next/server";


export async function POST(req: NextRequest, res: NextResponse) {
  try {
    const body = await req.json();
    console.log(body);  
    return NextResponse.json({ message: "Hello, world!" });
  } catch (err) {
    console.error(err);
    return NextResponse.error();
  }
}
```

And every time, the app crashes because the route handler fails with the following error:
```
- error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6571:19)
    at successSteps (node:internal/deps/undici/undici:6545:27)
    at node:internal/deps/undici/undici:1211:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

I'm usually inclined to think this is a silly user-error on my part, but I've seen similar issues in both this thread here, as well as [this issue that documents the same "problem" but for DELETE routes](https://github.com/vercel/next.js/issues/48096) (I put "problem" in quotations bc, as a poster in that thread notes, DELETE routes technically aren't supposed to support request bodies). 

A POST request should definitely support a body field in the request, though, and I can't figure out if I'm doing something wrong here or this is a Next issue

**Reply by pac008** [reply]
![image](https://github.com/vercel/next.js/assets/64404232/57b18d32-89c1-43a0-9464-f34a65782f34)

Check that you are sending a json. maybe that's the mistake

**Reply by pac008** [reply]
I had the same problem but it was because I was sending the data from "x-www-form-urlencoded"

**Reply by faaiz99** [reply]
I have the same error due to "x-www-form-urlencoded" how did you fix it? I cannot change the Content-Type, the documentation is not clear.

**Reply by pac008** [reply]
You must send a JSON, in the image above I show for example in postman that I put "raw", to then select "JSON". Then in your app you should see verify that you are sending a JSON, it is usually the default case in the "fetch".

**Reply by faaiz99** [reply]
That fixed it. Thanks!

**Reply by izzypt** [reply]
I was having the same issue and it is because I was using form-data in POSTMAN. Not sending the JSON in the raw , like 1st picture showed.

**Comment by donhamalainen** [comment]
So when I will use NextResponse and NextRequest ?

**Reply by pac008** [reply]
Well, I use the nextResponse to respond and the nextRequest as type. I don't know if I answer your question, I just don't know if I understood it. all the best!

**Comment by demirtasdurmus** [comment]
I understand that when we send a json body as a part of POST or PUT request, `req.json()` method works properly. But when the client payload is not json or does not exist at all; `req.json()` block throws a syntax error which is not a good way to handle if request data exists. In the end we can't know what a client will send, right?

`SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6662:19)
    at successSteps (node:internal/deps/undici/undici:6636:27)
    at node:internal/deps/undici/undici:1236:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)`

I think somehow we need a way to intercept the request object and check if data exists to send a 4.. error not a 500 error as it will happen otherwise?

**Comment by zeelz** [comment]
After battling with this issue for an hour, I had to comment `output` prop in `next.config.js` to get request.json() to deliver the request's body.

```
const nextConfig = {
    // "output": "export",
    "distDir": "build",
    experimental: {
        serverActions: true
    }
}
```

I'm still trying to understand the connection. Shouldn't I be able to build my client components to a dir?

**Reply by rizvan555** [reply]
Thank you very much 👍

**Comment by xmaxlad** [comment]
Hey, I am trying to shift my sever to nextjs but for the very basic route i can not access the req.body - 

This is what i am doing(The other routes with not tryin to access the body works fine)

Nextjs version - 14.2.15

my_root/app/api/courses/via-api/route.ts

import {NextResponse,NextRequest} from 'next/server'

export async function GET(req: Request,context:any) {

    try{
        const body = await req.json() 
    }catch(error){
        console.log(error)
        return NextResponse.json({
            message: "Error",
            error
        })
    }


    return NextResponse.json({})
}

Here is how i am trying to access the route in postman desktop app - 

GET/http://localhost:3000/api/courses/via-api

Body - raw - JSON - 
{
    "name":"kamal",
    "what":"backend"
}

But I am getting this error again and again - 
SyntaxError: Unexpected end of JSON input

**Reply by icyJoseph** [reply]
That looks like an issue with the POSTMAN request, try to do, `req.text()`, instead and see what that gives you back.
